File: src/main/java/xyz/phanta/cbtweaker/multiblock/MultiBlockData.java
package xyz.phanta.cbtweaker.multiblock;

import io.github.phantamanta44.libnine.util.data.ByteUtils;
import io.github.phantamanta44.libnine.util.data.ISerializable;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.recipe.RecipeExecutor;
import xyz.phanta.cbtweaker.recipe.RecipeLogic;
import xyz.phanta.cbtweaker.structure.StructureMatch;
import xyz.phanta.cbtweaker.util.RefreshState;
import xyz.phanta.cbtweaker.util.block.RedstoneControlHandler;
import xyz.phanta.cbtweaker.world.RoiHost;
import xyz.phanta.cbtweaker.world.RoiTicket;

import javax.annotation.Nullable;

public class MultiBlockData<R, C, D, S, E> implements RoiHost, ISerializable {

    public static <R, C, D> MultiBlockData<R, C, D, ?, ?> construct(MultiBlockControllerTileEntity mbCtrl,
                                                                    MultiBlockType<R, C, D> mbType) {
        return new MultiBlockData<>(mbCtrl, mbType, mbType.getRecipeLogic());
    }

    private final MultiBlockControllerTileEntity mbCtrl;
    private final MultiBlockType<R, C, D> mbType;
    private final RecipeLogic<R, C, D, S, E> recipeLogic;
    private final RedstoneControlHandler rsHandler = new RedstoneControlHandler(this::markControllerStateDirty);

    @Nullable
    private RoiTicket mbRoiTicket = null;
    @Nullable
    private MultiBlockAssembly<R, C, D, S, E> assembly = null;
    @Nullable
    private NBTTagCompound bufferedAssemblyDeser = null;
    private boolean structDirty = false;
    private RefreshState refreshState = RefreshState.NONE;

    private boolean wasActiveLastTick = false;

    public MultiBlockData(MultiBlockControllerTileEntity mbCtrl,
                          MultiBlockType<R, C, D> mbType, RecipeLogic<R, C, D, S, E> recipeLogic) {
        this.mbCtrl = mbCtrl;
        this.mbType = mbType;
        this.recipeLogic = recipeLogic;
        rsHandler.setRedstoneState(mbCtrl.getWorld().isBlockPowered(mbCtrl.getPos()));
    }

    public MultiBlockType<R, C, D> getMultiBlockType() {
        return mbType;
    }

    public RecipeLogic<R, C, D, S, E> getRecipeLogic() {
        return recipeLogic;
    }

    public boolean isAssembled() {
        return assembly != null;
    }

    public boolean isActive() {
        return rsHandler.canWork() && assembly != null && assembly.isActive();
    }

    public RedstoneControlHandler getRedstoneHandler() {
        return rsHandler;
    }

    public UiElement createRecipeLogicUiElement() {
        return recipeLogic.createUiElement(mbType.getRecipeConfig(),
                new RecipeExecutor.OfNullable<>(() -> assembly != null ? assembly.getRecipeExecutor() : null));
    }

    @Override
    public boolean isValidRoiHost() {
        return !mbCtrl.isInvalid();
    }

    private void dropRoi(boolean invalidateAssembly) {
        if (mbRoiTicket != null) {
            mbRoiTicket.invalidateRoi();
            if (invalidateAssembly) {
                if (assembly != null) {
                    assembly.disassociateHatches(mbCtrl);
                }
                structDirty = true;
            }
            mbRoiTicket = null;
        }
    }

    private void acquireBaseRoi() {
        World world = mbCtrl.getWorld();
        BlockPos pos = mbCtrl.getPos();
        mbRoiTicket = CbtMod.PROXY.getRoiTracker().registerRoi(this, world, mbType.getStructureMatcher()
                .getRegion(world, pos, world.getBlockState(pos).getValue(MultiBlockControllerBlock.PROP_DIRECTION)));
    }

    public void onInvalidated() {
        dropRoi(true);
    }

    @Override
    public void onRegionChanged(RoiTicket ticket, BlockPos pos) {
        Vec3i posOffset = pos.subtract(mbCtrl.getPos());
        if (posOffset.equals(Vec3i.NULL_VECTOR)) {
            dropRoi(true);
        }
        structDirty = true;
    }

    public void notifyHatchChanged(boolean compsDirty) {
        refreshState = compsDirty ? RefreshState.HARD_REFRESH : RefreshState.SOFT_REFRESH;
    }

    public void markControllerStateDirty() {
        mbCtrl.setDirty();
    }

    public void tick() {
        World world = mbCtrl.getWorld();
        BlockPos pos = mbCtrl.getPos();
        if (mbRoiTicket == null) {
            acquireBaseRoi();
            structDirty = true;
        }
        if (structDirty) {
            StructureMatch structMatch = mbType.getStructureMatcher()
                    .findMatch(world, pos, world.getBlockState(pos).getValue(MultiBlockControllerBlock.PROP_DIRECTION));
            if (structMatch != null) {
                assembly = MultiBlockAssembly.fromStructure(this, structMatch);
                // this only really works for the simple structure matcher, because other structures can potentially
                // grow without needing an original block removed, e.g. the linear matcher
                // in the future, could allow the matcher itself to define dynamic ROIs, but this works for now
                /*dropRoi(false);
                mbRoiTicket = CbtMod.PROXY.getRoiTracker()
                        .registerRoi(this, world, assembly.getStructureBlocks().iterator());*/
                assembly.associateHatches(mbCtrl);
                if (bufferedAssemblyDeser != null) {
                    assembly.deserializeFromNbt(bufferedAssemblyDeser);
                    bufferedAssemblyDeser = null;
                }
                // recheck recipe immediately once assembled
                // don't need to hard-refresh because the executor state will be fresh anyways
                refreshState = RefreshState.SOFT_REFRESH;
            } else {
                assembly = null;
                // not needed if the dynamic ROIs as mentioned above are not implemented
                //dropRoi(true);
            }
            structDirty = false;
        }
        boolean active;
        if (assembly != null && rsHandler.canWork()) {
            if (refreshState != RefreshState.NONE) {
                assembly.refreshCraftingState(refreshState == RefreshState.HARD_REFRESH);
                refreshState = RefreshState.NONE;
            }
            if (!world.isRemote) {
                assembly.tick();
            }
            active = assembly.isActive();
        } else {
            active = false;
        }
        if (wasActiveLastTick != active) {
            wasActiveLastTick = active;
            world.markBlockRangeForRenderUpdate(pos, pos);
        }
    }

    @Override
    public void serBytes(ByteUtils.Writer data) {
        if (assembly == null) {
            data.writeByte((byte)0);
        } else {
            assembly.serializeToBytes(data);
        }
        rsHandler.serBytes(data);
    }

    @Override
    public void deserBytes(ByteUtils.Reader data) {
        byte flag = data.readByte();
        if (flag != 0) {
            if (assembly == null) {
                return; // the rest of the packet will be unparsable if the assembly part isn't parsed out
            }
            assembly.deserializeFromBytes(data, flag);
        }
        rsHandler.deserBytes(data);
    }

    @Override
    public void serNBT(NBTTagCompound tag) {
        if (assembly != null) {
            assembly.serializeToNbt(tag);
        }
        rsHandler.writeToNbt(tag, "RedstoneControl");
    }

    @Override
    public void deserNBT(NBTTagCompound tag) {
        if (!tag.isEmpty()) {
            if (assembly != null) {
                assembly.deserializeFromNbt(tag);
            } else {
                bufferedAssemblyDeser = tag;
            }
        }
        rsHandler.readFromNbt(tag, "RedstoneControl");
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/multiblock/MultiBlockData.kt
package st.evening.mc.cbtweaker.multiblock

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.common.BufferedSyncHolder
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.serconfig.CopiableConfigHost
import st.evening.mc.cbtweaker.structure.StructureHighlightParticle
import st.evening.mc.cbtweaker.structure.StructureMatch
import st.evening.mc.cbtweaker.structure.StructureMatcher
import st.evening.mc.cbtweaker.structure.StructureVisualization
import st.evening.mc.cbtweaker.util.component.RedstoneControlHandler
import st.evening.mc.cbtweaker.util.machine.RefreshState
import st.evening.mc.cbtweaker.world.RoiHost
import st.evening.mc.cbtweaker.world.RoiTicket
import st.evening.mc.prelude.api.block.prefab.BlockSidedIfc
import st.evening.mc.prelude.api.data.ser.ServerSideSerializable
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.ServerSide
import st.evening.mc.prelude.api.util.game.onPhysicalClient
import st.evening.mc.prelude.api.util.world.onClient
import st.evening.mc.prelude.api.util.world.onServer

class MultiBlockData<S, M>(
    val mbCtrl: MultiBlockControllerTileEntity,
    val mbType: MultiBlockType<S>,
    private val structureMatcher: StructureMatcher<M>
) : RoiHost, CopiableConfigHost, BufferedSyncHolder, ServerSideSerializable {
    private var mbRoiTicket: RoiTicket? = null
    private var prevMatchData: M? = null
    private val changedBlocks: MutableSet<BlockPos> = mutableSetOf()
    private var structDirty: Boolean = true

    private var assembly: MultiBlockAssembly<S>? = null
    @ServerSide
    private var bufferedAssemblyDeser: NBTTagCompound? = null
    @ClientSide
    private var bufferedAssemblyBind: BindData? = null
    @ServerSide
    private var cachedAssemblyConfig: NBTTagCompound? = null
    var assemblyStateClock: Int = 0
        private set

    private var refreshState: RefreshState = RefreshState.NONE
    private var wasActiveLastTick: Boolean = false

    val assembled: Boolean
        get() = assembly != null

    val isActive: Boolean
        get() = assembly?.isActive == true

    val rsHandler: RedstoneControlHandler?
        get() = assembly?.rsHandler

    override val isValidRoiHost: Boolean
        get() = !mbCtrl.isInvalid

    private fun dropRoi() {
        mbRoiTicket?.let { ticket ->
            ticket.invalidateRoi()
            mbRoiTicket = null
        }
    }

    private fun dropAssembly() {
        assembly?.let {
            mbCtrl.world.onServer {
                val configDto = NBTTagCompound()
                it.writeConfig(configDto)
                if (!configDto.isEmpty) {
                    cachedAssemblyConfig = configDto
                }
                it.handleDestruction(mbCtrl.world.getBlockState(mbCtrl.pos))
            }
            it.disassociateHatches(mbCtrl)
            it.invalidate()
            assembly = null
            assemblyStateClock++
            mbCtrl.onAssemblyChanged(null)
        }
    }

    fun onInvalidated() {
        dropRoi()
        dropAssembly()
        changedBlocks.clear()
        structDirty = true
    }

    override fun onRegionChanged(ticket: RoiTicket, pos: BlockPos) {
        if (mbCtrl.isInvalid) return
        if (pos == mbCtrl.pos) { // either rotated or broken; either way, we'll need to re-check the entire structure
            onInvalidated()
        } else {
            changedBlocks += pos
            structDirty = true
        }
    }

    fun notifyHatchChanged(compsDirty: Boolean) {
        refreshState = if (compsDirty) RefreshState.HARD_REFRESH else RefreshState.SOFT_REFRESH
    }

    fun tick() {
        val world = mbCtrl.world
        val pos = mbCtrl.pos
        if (structDirty) {
            val match = structureMatcher.findMatch(
                world,
                pos,
                world.getBlockState(pos).getValue(BlockSidedIfc.PROP_FACING),
                prevMatchData,
                if (mbRoiTicket != null) changedBlocks else emptySet()
            )
            changedBlocks.clear()
            if (match is StructureMatch.Success<M>) {
                assembly?.invalidate()
                val assembly = MultiBlockAssembly.fromStructure(this, match.parts, assembly)
                this.assembly = assembly
                assemblyStateClock++
                mbCtrl.world.onClient {
                    bufferedAssemblyBind?.let {
                        assembly.bindSync(it.hostId, it.data)
                        it.data.release()
                        bufferedAssemblyBind = null
                    }
                    onPhysicalClient {
                        val fx = Minecraft.getMinecraft().effectRenderer
                        match.parts.positions.forEach {
                            fx.addEffect(StructureHighlightParticle(world, it, 16))
                        }
                    }
                }
                mbCtrl.onAssemblyChanged(assembly)
                assembly.associateHatches(mbCtrl)
                mbCtrl.world.onServer {
                    bufferedAssemblyDeser?.let {
                        assembly.readFromNbtServerSide(it)
                        bufferedAssemblyDeser = null
                    }
                    cachedAssemblyConfig?.let {
                        assembly.readConfig(it)
                        cachedAssemblyConfig = null
                    }
                }
                // recheck recipe immediately once assembled
                // don't need to hard-refresh because the executor state will be fresh anyways
                refreshState = RefreshState.SOFT_REFRESH
            } else {
                dropAssembly()
            }
            prevMatchData = match.data
            match.newRegion?.let {
                dropRoi()
                mbRoiTicket = CbTweaker.defns.roiTracker.registerRoi(this, world, it.iterator())
            }
            structDirty = false
        }

        val active: Boolean
        val assembly = this.assembly
        if (assembly != null) {
            if (refreshState != RefreshState.NONE) {
                assembly.refreshState(refreshState == RefreshState.HARD_REFRESH)
                refreshState = RefreshState.NONE
            }
            world.onServer {
                assembly.tick()
            }
            active = assembly.isActive
        } else {
            active = false
        }

        if (wasActiveLastTick != active) {
            wasActiveLastTick = active
            world.markBlockRangeForRenderUpdate(pos, pos)
        }
    }

    fun handleInteraction(
        state: IBlockState, player: EntityPlayer, hand: EnumHand,
        face: EnumFacing, hitX: Float, hitY: Float, hitZ: Float
    ): Boolean = assembly?.handleInteraction(state, player, hand, face, hitX, hitY, hitZ) ?: false

    @ServerSide
    fun handleBlockUpdate(state: IBlockState, fromBlock: Block, fromPos: BlockPos) {
        assembly?.handleBlockUpdate(state, fromBlock, fromPos)
    }

    @ServerSide
    fun handleDestruction(state: IBlockState) {
        assembly?.handleDestruction(state)
    }

    @ServerSide
    override fun writeConfig(dto: NBTTagCompound) {
        assembly?.writeConfig(dto)
    }

    @ServerSide
    override fun readConfig(dto: NBTTagCompound) {
        assembly?.readConfig(dto)
    }

    @ClientSide
    fun bindSync(hostId: Int, syncData: PacketBuffer) {
        val assembly = this.assembly
        if (assembly != null) {
            assembly.bindSync(hostId, syncData)
        } else {
            bufferSyncData(hostId, syncData)
        }
    }

    @ClientSide
    override fun bufferSyncData(hostId: Int, syncData: PacketBuffer) {
        bufferedAssemblyBind?.data?.release()
        syncData.retain()
        bufferedAssemblyBind = BindData(hostId, syncData)
    }

    @ServerSide
    override fun writeToNbtServerSide(dto: NBTTagCompound) {
        assembly?.writeToNbtServerSide(dto)
    }

    @ServerSide
    override fun readFromNbtServerSide(dto: NBTTagCompound) {
        if (!dto.isEmpty) {
            val assembly = this.assembly
            if (assembly != null) {
                assembly.readFromNbtServerSide(dto)
            } else {
                bufferedAssemblyDeser = dto
            }
        }
    }

    fun createMachineUiElement(): UiElement? = assembly?.createMachineUiElement()

    fun getStructureVisualization(): StructureVisualization = structureMatcher.getVisualization(prevMatchData)

    private class BindData(val hostId: Int, val data: PacketBuffer)
}
